src/sage/calculus/expr.py

```diff
@@ -0,0 +1,203 @@
+        [1 2 3]
+        [4 5 6]
+        sage: B.parent()
+        Full MatrixSpace of 2 by 3 dense matrices over Symbolic Ring
+
+    If ``x`` is a function, for example defined by a ``lambda`` expression, create a
+    symbolic function::
+
+        sage: f = symbolic_expression(lambda z: z^2 + 1); f
+        z |--> z^2 + 1
+        sage: f.parent()
+        Callable function ring with argument z
+        sage: f(7)
+        50
+
+    If ``x`` is a list or tuple of functions, or if ``x`` is a function that returns a list
+    or tuple, create a callable symbolic vector::
+
+        sage: symbolic_expression([lambda mu, nu: mu^2 + nu^2, lambda mu, nu: mu^2 - nu^2])
+        (mu, nu) |--> (mu^2 + nu^2, mu^2 - nu^2)
+        sage: f = symbolic_expression(lambda uwu: [1, uwu, uwu^2]); f
+        uwu |--> (1, uwu, uwu^2)
+        sage: f.parent()
+        Vector space of dimension 3 over Callable function ring with argument uwu
+        sage: f(5)
+        (1, 5, 25)
+        sage: f(5).parent()
+        Vector space of dimension 3 over Symbolic Ring
+
+    TESTS:
+
+    Lists, tuples, and vectors of length 0 become vectors over a symbolic ring::
+
+        sage: symbolic_expression([]).parent()
+        Vector space of dimension 0 over Symbolic Ring
+        sage: symbolic_expression(()).parent()
+        Vector space of dimension 0 over Symbolic Ring
+        sage: symbolic_expression(vector(QQ, 0)).parent()
+        Vector space of dimension 0 over Symbolic Ring
+
+    If a matrix has dimension 0, the result is still a matrix over a symbolic ring::
+
+        sage: symbolic_expression(matrix(QQ, 2, 0)).parent()
+        Full MatrixSpace of 2 by 0 dense matrices over Symbolic Ring
+        sage: symbolic_expression(matrix(QQ, 0, 3)).parent()
+        Full MatrixSpace of 0 by 3 dense matrices over Symbolic Ring
+
+    Also functions defined using ``def`` can be used, but we do not advertise it as a use case::
+
+        sage: def sos(x, y):
+        ....:     return x^2 + y^2
+        sage: symbolic_expression(sos)
+        (x, y) |--> x^2 + y^2
+
+    Functions that take a varying number of arguments or keyword-only arguments are not accepted::
+
+        sage: def variadic(x, *y):
+        ....:     return x
+        sage: symbolic_expression(variadic)
+        Traceback (most recent call last):
+        ...
+        TypeError: unable to convert <function variadic at 0x...> to a symbolic expression
+
+        sage: def function_with_keyword_only_arg(x, *, sign=1):
+        ....:     return sign * x
+        sage: symbolic_expression(function_with_keyword_only_arg)
+        Traceback (most recent call last):
+        ...
+        TypeError: unable to convert <function function_with_keyword_only_arg at 0x...>
+        to a symbolic expression
+    """
+    if isinstance(x, Expression):
+        return x
+    elif hasattr(x, '_symbolic_'):
+        return x._symbolic_(SR)
+    elif isinstance(x, (tuple, list, FreeModuleElement)):
```

Comment:
I'm curious, why move away from using `is_FreeModuleElement` but keep using `is_Matrix` (line 181)?

Reply:
Could as well also replace `is_Matrix`, but that is defined (along with the ABC `Matrix`) in the depths of `sage.structure.element`, which is shipped already with the lowest-level distribution, **sagemath-objects**, so it is not relevant for the modularization.

Reply:
... also note that in the systematic deprecation of `is_...` functions, I have taken care of parent classes primarily so far.

Reply:
Thanks for the explanation, that makes sense.